WebMIDI In/Out extension

lselden opened the pull request:
This is an extension that adds MIDI support using the WebMIDI API. It tries to cover most use cases - it supports notes, CC, Pitch Bend, Program Change messages, etc.

## Features

* Device detection/enumeration
* MIDI Input (When Note On/Off, When any midi event)
* MIDI Output (similar syntax to the "Music" extension) - uses music "tempo" if set
* Event reporter for thread context's current event
* human-readable (and forgiving) string formatting of events for parsing/formatting

## Todo

- [ ]  Translations
- [ ]  Add blocks documentation
- [ ]  Add string format documentation
- [ ]  Add "Getting Started" / browser compatibility documentation
- [ ]  Test on target devices (in particular mac / android)
- [ ]  Feedback on blocks - are any missing? should some be pruned? Better naming convention?
- [ ]  Image
- [ ]  (OPTIONAL) Separate icons for input vs output?

lselden committed:
Add midi extension

Added midi extension for input/output of midi using the WebMidi API. Still pending documentation and peer-review

lselden committed:
Add midi extension

Added midi extension for input/output of midi using the WebMidi API. Still pending documentation and peer-review

lselden committed:
updated midi.js - tweak string/parsing/formatting

use "t=<time>" instead of "@<time>". Allow using "type=note pitch=60" instead of "note 60" if need to be verbose. Some bugfixes

lselden committed:
updated midi.js - tweak string/parsing/formatting

use "t=<time>" instead of "@<time>". Allow using "type=note pitch=60" instead of "note 60" if need to be verbose. Some bugfixes

lselden committed:
midi - cleanup comments and types

lselden committed:
midi - cleanup comments and types

lselden committed:
Update midi.js

lselden committed:
Update midi.js

lselden committed:
midi.js - linting and ensure all inputs use Scratch.Cast

lselden committed:
midi.js - linting and ensure all inputs use Scratch.Cast

lselden committed:
Update midi.js

lselden committed:
Update midi.js

Brackets-Coder commented:
At a glance, this looks much better than mine. You probably spent a lot more than four hours on it like I did... I might give it a full review (just for my opinion) when I have more time but I'm a bit busy right now. If you'd like, I can submit a PR or so to your fork that adds translations so you have one less thing to worry about (but again, I'm a bit busy right now, I'll have to get around to it when I can).

lselden committed:
midi - add "whenDeviceEvent" hat for detecting device connect/disconnect

lselden committed:
midi - add "whenDeviceEvent" hat for detecting device connect/disconnect

lselden committed:
add midi to extensions list

lselden committed:
add midi to extensions list

Brackets-Coder commented on extensions/midi/midi.js:
```diff
@@ -0,0 +1,2284 @@
+// Name: MIDI
+// ID: extmidi
```
This should contain your username, generally matching the folder structure as mentioned above

Brackets-Coder commented on extensions/midi/midi.js:
```diff
@@ -0,0 +1,2284 @@
+        typeof BEATS === "string" && BEATS.includes("/")
+          ? parseFraction(BEATS)
+          : Cast.toNumber(BEATS);
+      // sanity check - don't let a note be more than 1 minute
+      // if anyone ever needs a longer note then they should manually write event string
```
Why?
I know it's a rare occurrence but as a musician (pianist), developer, and Scratch user you shouldn't generally limit the user's capabilities unless necessary. Notes technically _can_ be a minute long and a user shouldn't have that length unexpectedly cut off...
Keep in mind this is just a quick look-over and I haven't fully analyzed it in depth yet

lselden replied:
I hear ya. I checked and the Scratch Music extension clamps to 100 beats max (100 seconds at 60bpm) - updated code accordingly

yuri-kiss replied:
Following what the music extension does isnt required but you can if you want

Brackets-Coder commented on extensions/midi/midi.js:
```diff
@@ -0,0 +1,2284 @@
+      // sanity check - don't let a note be more than 1 minute
+      // if anyone ever needs a longer note then they should manually write event string
+      dur = Math.max(0, Math.min(dur, 60));
+      // convert from percent to midi value
+      let velocity = (Cast.toNumber(VELOCITY) / 100) * 127;
+      // clamp value
+      // NOTE - if 0 velocity then nothing will happen because treated as note off
+      velocity = Math.max(0, Math.min(velocity, 127));
+      // REVIEW - should OUTPUT_DEVICES be changed to output device index instead of id?
+      const deviceId = Cast.toString(DEVICE);
+      const device = this.midi.outputs.findIndex((d) => d.id === deviceId);
+      /** @type {MidiEvent} */
+      const event = {
+        type: "noteOn",
+        pitch: this.noteForName({ NOTE }),
+        velocity,
+        channel: Cast.toNumber(CHANNEL) || undefined,
```
casting here too

Brackets-Coder commented on extensions/midi/midi.js:
```diff
@@ -0,0 +1,2284 @@
+      }
+
+      // clamp values
+      let value1 = Cast.toNumber(VALUE1);
+      // pitchbend can be a bigger number
+      const maxValue1 = spec?.highResParam ? 16384 : 127;
+      value1 = Math.max(0, Math.min(value1, maxValue1));
+
+      let value2 = Cast.toNumber(VALUE2);
+      value2 = Math.max(0, Math.min(value2, 127));
+
+      // REVIEW - should OUTPUT_DEVICES be changed to output device index instead of id?
+      const deviceId = Cast.toString(DEVICE);
```
More casting

Brackets-Coder commented on extensions/midi/midi.js:
```diff
@@ -0,0 +1,2284 @@
+      return midiToString(event);
+    }
+    normalizeMidiVal({ VALUE, MIN, MAX }) {
+      const min = Cast.toNumber(MIN);
+      const max = Cast.toNumber(MAX);
+      const val = Cast.toNumber(VALUE) / 127;
```
You know what I'm going to say: `Scratch.Cast`

Brackets-Coder commented on extensions/midi/midi.js:
```diff
@@ -0,0 +1,2284 @@
+  }
+
+  //#endregion
+  Scratch.extensions.register(new MidiExtension());
```
This class name is a bit overgeneralized?
Also your code style is not exactly like mine but I guess it's just a matter of preference. seems a bit cluttered to me but I guess it's just a personal problem.

Other than that it looks fine

yuri-kiss replied:
The extension class doesnt have to be anything special

Brackets-Coder commented:
Just a quick review, nothing in depth. Going over best practices for TW extensions in particular, not necessarily anything related to the functionality of the code.

lselden committed:
bugfix in string parsing and minor formatting

lselden committed:
bugfix in string parsing and minor formatting

lselden committed:
midi - change params for sendOutputEvent

lselden committed:
midi - change params for sendOutputEvent

lselden committed:
midi - treat "pos" and "beats" as beats unit, "time" and "dur" as seconds

lselden committed:
midi - treat "pos" and "beats" as beats unit, "time" and "dur" as seconds

lselden committed:
allow string event input into "play note for beats" command

lselden committed:
allow string event input into "play note for beats" command

lselden commented on extensions/midi/midi.js:
```diff
@@ -0,0 +1,2284 @@
+   */
+  function noteNameToMidiPitch(note, defaultOctave = 4) {
+    const parts =
+      /(?<pitch>[A-G])(?<flat>[b♭]+)?(?<sharp>[#♯]+)?_?(?<octave>-?\d+)?/i.exec(
```
It's a regex with named groups to describe what's matched. added method comments to help describe what's valid. If anyone's interested in other implementations I used these for reference: [strudel](https://github.com/tidalcycles/strudel/blob/main/packages/core/util.mjs#L27) and [tonaljs](https://github.com/tonaljs/tonal/blob/main/packages/pitch-note/index.ts#L83)

lselden commented on extensions/midi/midi.js:
```diff
@@ -0,0 +1,2284 @@
+  function stringToMidi(text, opts = {}) {
+    if (typeof text !== "string") text = Cast.toString(text);
+    const fullRe =
+      /^\s*(?<type>[a-zA-Z]{2,})?\s*((?<pitch>[A-G][#b♯♭_]*-?\d?)|(?<data1>\b-?[0-9a-f]{1,5}\b))?\s*(?<data2>\b[0-9a-f]{1,3}\b)?\s*(?<keyvals>.+)\s*$/;
```
`Cast` here is a local const of `Scratch.Cast`. Switched to the verbose `Scratch.Cast` to avoid confusion in the future

yuri-kiss replied:
> `Cast` here is a local const of `Scratch.Cast`. Switched to the verbose `Scratch.Cast` to avoid confusion in the future

This can be changed back, using the `const` is fine, and is more readable

lselden commented on extensions/extensions.json:
```diff
@@ -48,6 +48,7 @@
   "ar",
   "encoding",
   "Lily/SoundExpanded",
+  "midi/midi",
```
"You can name your folder anything you want" - I'd rather keep it generic than use my username.

lselden commented on extensions/midi/midi.js:
```diff
@@ -0,0 +1,2284 @@
+          DEVICE_PROP: {
+            acceptReporters: true,
+            items: [
+              { value: "id", text: Scratch.translate("Device ID") },
+              {
+                value: "manufacturer",
+                text: Scratch.translate("Manufacturer"),
+              },
+              { value: "name", text: Scratch.translate("Device Name") },
+              { value: "state", text: Scratch.translate("State") },
```
fair enough - that's prettier's formatting choices, but made everything multiline instead.

lselden committed:
reformatting as per comments

lselden committed:
reformatting as per comments

lselden committed:
ensure device index starts at 1

lselden committed:
ensure device index starts at 1

lselden committed:
Merge remote-tracking branch 'upstream/master' into ext_midi

lselden committed:
Merge in MIDI Parser extension by CHCAT1320

lselden committed:
Merge in MIDI Parser extension by CHCAT1320

lselden committed:
minor refactor - rearrange code

lselden committed:
minor refactor - rearrange code

lselden committed:
change parseMidiDataUrl to use same keys

Existing code uses "dur" for "duration", and "pitch" for "noteName"

lselden committed:
change parseMidiDataUrl to use same keys

Existing code uses "dur" for "duration", and "pitch" for "noteName"

lselden committed:
round midi file times to 1 millisecond

midi precision doesn't go under 2-5 milliseconds, so forcing numbers to be readable 1ms precision

lselden committed:
round midi file times to 1 millisecond

midi precision doesn't go under 2-5 milliseconds, so forcing numbers to be readable 1ms precision

lselden committed:
fix json to event formatting

lselden committed:
fix json to event formatting

lselden committed:
fix linting

lselden committed:
fix linting

lselden committed:
Merge branch 'ext_midi' of https://github.com/lselden/extensions into ext_midi

lselden committed:
allow formatting pitch as sharp/flat or number

lselden committed:
always include node/noteon type in string output and catch some edge input cases

lselden committed:
add generic key=value parser/formatter

lselden committed:
revamp midi file parsing to use Scratch.fetch, include meta events like tempo/keySignature

lselden committed:
add license credits

lselden committed:
add midi extension to extensions.json

lselden committed:
change "noteOn" to "note", add Duration/Instrument event props, bugfixes

AliceC-W commented:
Hello,

This doesn't seem to read my MIDI devices after my app has packaged (at least on my Android when packaged from Penguinmod)

(To play a note on this, swipe from one clock position toward another (I will be adding more features to the face of this "clock, such as aftertouch, sustain, page turning, and menu))

Thanks for any help you can provide


[Flick Talk.zip](https://github.com/user-attachments/files/21081674/Flick.Talk.zip)

lselden commented:
> Hello,
> 
> This doesn't seem to read my MIDI devices after my app has packaged (at least in Chrome on my Android when packaged from Penguinmod)
> 
> (To play a note on this, swipe from one clock position toward another (I will be adding more features to the face of this "clock, such as aftertouch, sustain, page turning, and menu))
> 
> Thanks for any help you can provide
> 
> [Flick Talk.zip](https://github.com/user-attachments/files/21081674/Flick.Talk.zip)

@AliceC-W I couldn't get your app to do anything, so can't comment on it besides it looked like it's trying to play a 0 duration note, which might be an issue. That being said, browsers are particular about allowing MIDI access depending on the context - are you trying to open it as a local file? If so then try running it from a https link instead, which has fewer restrictions. You may also want to test if midi playback is supported by using the `number of [output] devices` block - it'll be 0 if no detected midi devices are available.

If you want further help then please make an example that just has the minimum necessary to reproduce (for example click/touch on the screen to play a single note)

AliceC-W commented:
> > Hello,
> > This doesn't seem to read my MIDI devices after my app has packaged (at least in Chrome on my Android when packaged from Penguinmod)
> > (To play a note on this, swipe from one clock position toward another (I will be adding more features to the face of this "clock, such as aftertouch, sustain, page turning, and menu))
> > Thanks for any help you can provide
> > [Flick Talk.zip](https://github.com/user-attachments/files/21081674/Flick.Talk.zip)
> 
> @AliceC-W I couldn't get your app to do anything, so can't comment on it besides it looked like it's trying to play a 0 duration note, which might be an issue. That being said, browsers are particular about allowing MIDI access depending on the context - are you trying to open it as a local file? If so then try running it from a https link instead, which has fewer restrictions. You may also want to test if midi playback is supported by using the `number of [output] devices` block - it'll be 0 if no detected midi devices are available.
> 
> If you want further help then please make an example that just has the minimum necessary to reproduce (for example click/touch on the screen to play a single note)

Yes I am trying to run it right on my phone in Android Chrome
(my address bar says content://media/external/file/1002145829 (in case that helps))

How do I run it from a https link? Will this require additional hardware (counting codebeautify.org)? Does Penguinmod offer a way to force that change in the address bar?

Here is a simple project detailing my only problem at the moment

[Number of MIDI Devices  Project.zip](https://github.com/user-attachments/files/21150275/Number.of.MIDI.Devices.Project.zip)

lselden commented:
> Yes I am trying to run it right on my phone in Android Chrome (my address bar says content://media/external/file/1002145829 (in case that helps))
> 
> How do I run it from a https link? Will this require additional hardware (counting codebeautify.org)? Does Penguinmod offer a way to force that change in the address bar?
> 
> Here is a simple project detailing my only problem at the moment
> 
> [Number of MIDI Devices Project.zip](https://github.com/user-attachments/files/21150275/Number.of.MIDI.Devices.Project.zip)

Thanks for providing the project. I'm able to see my output devices on my laptop, and 0 devices on my iPhone (which is expected - Apple doesn't support WebMidi).

I'd recommend using Github Pages to host your content (which is free) if you don't mind it being publically accessible - I'm sure there's plenty of tutorials out there if you aren't that familiar with github. You could also try testing if you have a computer on the same network using any static web server software, for example [fenix](https://github.com/coreybutler/fenix/releases), `python -m http.server` or `npx http-server` *(if you have python3 / node.js installed - google it)*. However, you may need to be using a `https://` link ([secure context](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts/features_restricted_to_secure_contexts)) to get WebMidi to work - check to see if WebMidi works on the [HTTP version of permission.site](http://permission.site/) vs the [HTTPS version of permission.site](https://permission.site).

AliceC-W commented:
Thanks

I just found out that Opera GX recognizes my MIDI devices even in airplane mode

I think it will do, at least until that environment can be simulated in the .html (directory) itself

I'm sorry to hear that about iPhone, definitely let me know if that changes



The equipment I used was:
1) NUOSIYA Type-C to MIDI Cable 6.5Ft,Type-C MIDI Interface Adapter,MIDI to Type-C Cable Converter ​for Music Keyboard Piano to PC Mac Laptop Windows Android (purchased on Amazon.com on Jul 4, 2024)

2) Android 13 on Samsung Galaxy A52 5G (Purchase recieved Mon, Sep 20, 2021)

3) Yamaha PSR-275 (gifted on December 25, 2003 before I even knew that MIDI is actually just an interface)

Brackets-Coder commented:
@lselden I've used both macOS and Windows throughout various periods. About the iPhone -- It's not that Apple straight up just doesn't support WebMIDI, it's that all Safari Browsers across Apple devices don't support the API due to security concerns, which Apple seems to care a _lot_ about. I've gotten WebMIDI to work as expected with Google Chrome on macOS,

lselden committed:
Merge remote-tracking branch 'upstream/master' into ext_midi

lselden committed:
remove incomplete mid file support

extension had a WIP addition of the @tonejs/midi library for midi file support. Removing in favor of it being a separate extension

lselden commented:
@Brackets-Coder sure go ahead and review.

Brackets-Coder commented on extensions/midi/midi.js:
```diff
@@ -0,0 +1,2557 @@
+// Name: MIDI
+// ID: extmidi
```
Generally IDs are supposed to contain the author's username to differentiate it from generic titles (such as many extensions being called "midi" or "3d". @CubesterYT is this ID unique enough to pass or should it include something more?

CubesterYT replied:
I agree, ID's always require the creator's name in them, so `iseldenMidi` would be better

Brackets-Coder replied:
Do we want to commit that in?

Brackets-Coder commented on extensions/midi/midi.js:
```diff
@@ -0,0 +1,2557 @@
+// ID: extmidi
+// Description: An extension to use the WebMidi API for midi input/output.
+// By: lselden
+// By: CHCAT1320
```
Are there no scratch accounts that these link to?

lselden replied:
Nope - I don't have a scratch account

Brackets-Coder reviewed (changes requested):
@lselden I did a bit of a hasty review of the functionality of some (not all) of your blocks, let me know if I missed anything or am misunderstanding. Thanks!

- The icon for the block palette is illegible in dark mode, a simple white circle behind it should suffice.
- The MIDI input blocks I tested all seem to be working as expected.
- The MIDI [connected/disconnected] hat block did not activate a visual response when I powered on and off my Roland FP-90 (bluetooth MIDI). There were no errors in the console when this happened. I will likely have to test with a MIDI cable later and see if that fixes the issue
- The "send" midi output block returned the following error:
`Uncaught TypeError: can't access property "device", event is null`
My keyboard is capable of MIDI output and the "number of [output] devices" reported that the extension was aware of its presence.

The error occurred on line 2292:
```javascript
    sendOutputEvent({ EVENT, DEVICE }, util) {
      const text = Scratch.Cast.toString(EVENT);
      const event = stringToMidi(text);

      if (DEVICE) {
        event.device = this._getDeviceIndex(DEVICE, "output");
      }

      this.midi.sendOutputEvent(event);
    }
```

I don't know what was causing this issue, but it's likely the "stringToMidi" function that's not parsing something correctly (I haven't gone through it to check) I used the default input. 

I've got another thing I have to run off to and I think I forgot to say something that I tested but let me know if I can do anything more for you

lselden committed:
add white circle to midi menu icon

lselden committed:
bugfix - correct default value of send output event

lselden committed:
detect invalid midi out events

lselden commented:
> @lselden I did a bit of a hasty review of the functionality of some (not all) of your blocks, let me know if I missed anything or am misunderstanding. Thanks!
> 
> * The icon for the block palette is illegible in dark mode, a simple white circle behind it should suffice.

Good suggestion - fixed.

> * The MIDI [connected/disconnected] hat block did not activate a visual response when I powered on and off my Roland FP-90 (bluetooth MIDI). There were no errors in the console when this happened. I will likely have to test with a MIDI cable later and see if that fixes the issue

This may be a quirk of chromium-based browsers. I didn't get a disconnect event when I disconnected midi devices, only triggers when a new port/device is added. There wasn't any `'statechange'` dom event fired, so no visual change. You can compare against behavior on https://studiocode.dev/midi-monitor/

> * The "send" midi output block returned the following error: 
> I don't know what was causing this issue, but it's likely the "stringToMidi" function that's not parsing something correctly (I haven't gone through it to check) I used the default input.

The default input was wrong, it needed to indicate it was a note on event (`note`). Also I added a guard again so that invalid strings don't throw an error (but will log to console.debug)

Brackets-Coder commented:
!format

Brackets-Coder commented:
I'll see if I can get in another review this weekend
I'm a bit busy and only busier from here

DangoCat committed:
[Automated] Format code

Brackets-Coder commented:
Sorry I haven't reviewed yet, please remind me to do one this weekend

lselden commented:
reminder 😀

On Thu, Jan 15, 2026 at 8:22 PM Brackets-Coder ***@***.***>
wrote:

> *Brackets-Coder* left a comment (TurboWarp/extensions#2022)
> <https://github.com/TurboWarp/extensions/pull/2022#issuecomment-3757636915>
>
> Sorry I haven't reviewed yet, please remind me to do one this weekend
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/TurboWarp/extensions/pull/2022#issuecomment-3757636915>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAWFHIOBUZAZFEQ64J27QX34HA4MPAVCNFSM6AAAAABYWHYIGCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTONJXGYZTMOJRGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Brackets-Coder commented:
> reminder 😀

Okay thank you, I'll do it today

Brackets-Coder commented on extensions/midi/midi.js:
```diff
@@ -0,0 +1,2563 @@
+      i -= 1;
+    }
+    return i;
+  };
```
Will the 5% of browsers that don't have findLastIndex even be able to run Turbowarp or the MIDI API?

Brackets-Coder commented on extensions/midi/midi.js:
```diff
@@ -0,0 +1,2563 @@
+    NOTEANY: `${EXT_ID}_whenAnyNoteOnOff`,
+    MIDI: `${EXT_ID}_whenMidiEvent`,
+    // not currently implemented
+    // CC: `${EXT_ID}_whenCC`
```
Why not?

lselden replied:
The "When CC Event" block was to support targeting when events happen on a specific CC channel. But in my testing using the "When MIDI Event" hat with a switch/case pattern was noticeably more performant than having multiple hats listing for separate CC events. I removed the comments

Brackets-Coder commented on extensions/midi/midi.js:
```diff
@@ -0,0 +1,2563 @@
+      if (event && ("pitch" in event || "value" in event)) {
+        delete event.value1;
+        delete event.value2;
+      }
```
I don't know if you wanted me to look at this...

lselden replied:
nope!

Brackets-Coder commented on extensions/midi/midi.js:
```diff
@@ -0,0 +1,2563 @@
+        delete event.value2;
+      }
+      // QUESTION does this need an error handler?
+      return JSON.stringify(event);
```
If you're checking to see if `event` is valid above, you should probably check to see if it's valid here too and that should be enough error handling

lselden replied:
will be `"null"` if not valid which I think should be safe

Brackets-Coder reviewed (approved):
- [x] I verified that the extension follows all acceptance criteria and guidelines in CONTRIBUTING.md.
 - [x] I believe the block list is well-designed and intuitive.
 - [x] I tested all functionality and found no bugs, security flaws, or unintuitive behavior.
 - [x] I reviewed the code and believe it is well-written and easy to maintain or extend in the future, even without help from the original author.
 - [ ] I have no additional concerns.

I'm going to approve because this extension is "Good enough" for merging but there are still things I think can be improved

- The icon is better, nice job
- Note outputs work by default now, nice job
   - Kinda untopical but today was the first day I ever heard my keyboard's MIDI output     
   and it sounds kinda weird
- The text for the block menu should probably be MIDI instead of Midi, but this is insignificant
- I wasn't able to test everything as I don't have a keyboard with pitch bend, etc. but it does have a pedal

As @SharkPool-SP mentioned in #1974, most lists should be wrapped in JSON.stringify as that makes it easier to work with in other extensions. I don't think this is required for your extension to be acceptable since there's already a "get [index] of active notes" block, but it still might be nice if you made the active notes return a JSON array.
Also, it will definitely be helpful for users if you add documentation. One of the most valuable blocks is the one that separates MIDI events into the specified property and returns it, but to me it wasn't very clear initially that "(type eventType) of (note C4 82 ch1 dev0 @1000)" did this.

Documentation will definitely help a lot, and while it isn't required for this to be merged, I would definitely 100% recommend. Especially with the event formatting—it can definitely be confusing for users who don't understand what the properties correspond to and how to decode it.

I've done a slow skim of the code and it looks fine, but I haven't critically analyzed the nooks and crannies of every line.

Brackets-Coder commented:
@SharkPool-SP mind doing a quick review?
@lselden is there anything you know you want done prior to merging?

SharkPool-SP commented:
I don't have a midi to test with

Brackets-Coder commented:
> I don't have a midi to test with

That's right, I forgot you told me
I don't suppose any of the others do either...
Could you just do a quick code skim and see if I missed anything? You don't have to approve if you feel like you can't test

SharkPool-SP commented:
> > I don't have a midi to test with
> 
> That's right, I forgot you told me I don't suppose any of the others do either... Could you just do a quick code skim and see if I missed anything? You don't have to approve if you feel like you can't test

okay

SharkPool-SP commented:
ill do it when they fix the other requested changes

lselden committed:
midi - add json output for lists, cleanup comments

added extra block for outputting devices as json, allow outputting active notes as json

lselden committed:
Merge branch 'ext_midi' of https://github.com/lselden/extensions into ext_midi

lselden commented:
@SharkPool-SP I addressed comments and added some JSON output blocks. I also plan on adding some documentation.

SharkPool-SP commented:
Then I'll wait until documentation is made before merging

lselden committed:
parse midi string fixes - handle beats, "rest" type

lselden committed:
correctly clear input note buffer and allow passing full midi message to "is note active", not just the pitch

lselden committed:
Make Send Midi Output event match behavior of Music extension

return promise to suspend execution for duration of event

lselden committed:
midi - added documentation and sample file

lselden committed:
add input device selection to when hats

Brackets-Coder commented on docs/midi/MIDI.md:
```diff
@@ -0,0 +1,260 @@
+| `time` | `t` / `@` | number | Time of event in **seconds** |
+| `pos` | - | number | Time in **beats** (converted to time using current tempo) |
+| `dur` | `duration` | number | **[Output only]** Duration in **seconds** (for `note` type, sends corresponding MIDI note off message automatically) |
+| `beats` | - | number | **[Output only]** Duration in **beats** - (converted to `duration` using current tempo) |
```
These and the MIDI maapings below: while possibly helpful for advanced users, they may not be necessary for introductory documentation.

lselden replied:
Agreed, reasonable. I removed most of the appendix

Brackets-Coder commented on samples/Midi.sb3:
This file makes validation fail because you loaded the extension from localhost. I would suggest removing it until the extension is merged, then add the sample project using extensions.turbowarp.org in a new pull request.

lselden replied:
removed!

Brackets-Coder reviewed (changes requested):
This isn't a code review, I'm just looking over the documentation with two quick suggestions. Let me know if you have any questions or need anything. 

I haven't tested the sample project or code yet after my previous review, but everything else looks okay

lselden committed:
simplify midi docs, remove sample file
